Use phpunit groups to run database code in isolation

DeepDiver1975 opened the pull request:
There have already been some code pieces to look into:
- [ ] Why do unit tests for preview generation require database access? (Yes - filesystem foo - but could be structured differently)
- [ ] Pure tests on the server container - why? 
- [ ] DIContainerTest
- [ ] avatar tests
- [x] avatar controller tests #20510
- [ ] stream wrapper tests
- [ ] OCA\Provisioning_API\Tests\AppsTest needs mocks
- [ ] ShareControllerTest needs to mock file system
- [ ] All LDAP unit tests require DB access - even those who already mock IConfig etc :fish: -y
- [ ] Test_UrlGenerator - basically the UrlGenerator needs to inject the router object
- [ ] https://github.com/owncloud/core/issues/20771
- [ ] Test_Util_CheckServer on postgresql

TODO:
- review all unit test in group DB - is real database access really required here?
- review all still failing unit tests - is real database access required?
- think about more separation of existing classes to get more isolated testing (e.g. preview generation)

DeepDiver1975 committed:
A TestCase which is not annotated to be in group DB will not allow access to the database connection.
This is necessary to categorize unit test and avoid duplicate test case execution - it also allows us to closely review unit test implementations to identify unnecessary db calls.

DeepDiver1975 committed:
Introduce evaluation of an environment variable to choose if all, only database or only database unit tests are to be executed.
This is mainly required on automated execution on Jenkins

DeepDiver1975 committed:
Apply DB group annotation ...

rullzer commented:
>(Yes - filesystem foo - but could be structured differently)

Seems like a good time to revive https://github.com/owncloud/3rdparty/pull/185

MorrisJobke commented:
I also thought about this some time ago. I didn't know of this grouping. :)

MorrisJobke commented on autotest.sh:
```diff
@@ -260,14 +260,27 @@ function execute_tests {
 	if [[ "$_XDEBUG_CONFIG" ]]; then
 		export XDEBUG_CONFIG=$_XDEBUG_CONFIG
 	fi
+	if [ -z "$TEST_SELECTION" ]; then
```
Why is this needed? The variable is not used below.

DeepDiver1975 replied:
:see_no_evil: :question: :question:

MorrisJobke replied:
There is no need to change the value from empty to `all` because below only `DB` and `NODB` are checked. Why adding code that is useless and only confuses ;)

DeepDiver1975 replied:
> only confuses ;)

oh - it's that easy to confuse you :smiling_imp:

DeepDiver1975 committed:
Mock DIContainer to not hit the database

DeepDiver1975 committed:
Prevent redefinition warning of PHPUNIT_RUN

DeepDiver1975 committed:
Adding more test suites to group DB

DeepDiver1975 committed:
Further unit test suite tweats

DeepDiver1975 committed:
External storage tests require DB as well

DeepDiver1975 committed:
Adding more DB groups to test suits

Adding more DB groups to test suits

DeepDiver1975 committed:
Un-confuse @MorrisJobke :wink:

DeepDiver1975 committed:
Add DB group to some files_external tests

Adding group Db to federation tests and ldap tests

Add group DB to Test_UrlGenerator

Adding group DB to trashbin and versions tests

Adding group DB to Test_Util_CheckServer for pg

DeepDiver1975 commented:
@nickvergessen any clue why this in only on php7? THX

```
03:50:59 1) OC\Settings\Controller\EncryptionControllerTest::testStartMigrationSuccessful
03:50:59 OCP\AutoloadNotAllowedException: Autoload path not allowed: /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/apps/encryption/lib/migration.php
03:50:59 
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/autoloader.php:154
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/lib/autoloader.php:174
03:50:59 /ssd/jenkins/workspace/server-master-linux-php7-ci/database/sqlite/label/SLAVE/tests/settings/controller/EncryptionControllerTest.php:94

nickvergessen commented:
Well encryption app is not enabled by default, so it's not enabled, so the paths are not allowed to be loaded.

Why this only happens in php7? I think because 
```php
		$migration = $this->getMockBuilder('\\OCA\\Encryption\\Migration')
			->disableOriginalConstructor()->getMock();
```
uses Reflection class and that behaviour might have changed, due to type hinting.
So the fix is to add the apps/encryption manually to the loaded paths:
https://github.com/owncloud/core/blob/master/tests/bootstrap.php#L11

DeepDiver1975 commented:
> Well encryption app is not enabled by default, so it's not enabled, so the paths are not allowed to be loaded.

but we enable it in https://github.com/owncloud/core/blob/master/tests/enable_all.php

DeepDiver1975 commented:
> but we enable it in https://github.com/owncloud/core/blob/master/tests/enable_all.php

https://github.com/owncloud/core/blob/master/autotest.sh#L259

DeepDiver1975 committed:
Don't execute phpunit twice

DeepDiver1975 committed:
More group DB

DeepDiver1975 commented:
Let's get this in @MorrisJobke @PVince81 @icewind1991 @blizzz THX

nickvergessen commented on tests/lib/util.php:
```diff
@@ -167,55 +167,6 @@ function testGetInstanceIdGeneratesValidId() {
 	}
 
 	/**
-	 * Tests that the home storage is not wrapped when no quota exists.
```
huu? Is this intended?

DeepDiver1975 replied:
hmmmm - rebase-fu - let me double check

DeepDiver1975 replied:
Ah .. I moved them to a file of it's own: https://github.com/owncloud/core/commit/cb74b575afbf9a086e2eb4980ce45641e65a8bef#diff-17da1c724f4cabe27a25e2108e779f65R29

MorrisJobke replied:
But this commit is missing in this PR.

nickvergessen replied:
he just rebased afterwards again

nickvergessen replied:
https://github.com/owncloud/core/pull/20248/files#diff-17da1c724f4cabe27a25e2108e779f65

nickvergessen commented on tests/lib/testcase.php:
```diff
@@ -49,6 +51,16 @@ protected function getTestTraits() {
 
 	protected function setUp() {
+		// detect database access
+		if (!$this->IsDatabaseAccessAllowed()) {
```
This will break all apps I think, not sure how to do it better,
so please drop a mail to the mailing list asap?

DeepDiver1975 replied:
> so please drop a mail to the mailing list asap?

Or we disable this for apps running on travis - we need this for core only from my pov

nickvergessen replied:
well it can help in apps as well, only some tests require the DB and then they can make use of it.

About disabling for apps, the tests should pass locally as well, so the only way I could imagine is checking the class namespace: if start with OCA and not with OCA\Dav|Files|...\ then ignore

DeepDiver1975 replied:
> well it can help in apps as well, only some tests require the DB and then they can make use of it.

sure - but I'd take care of this in a PR of it's own - let's just make sure this PR does not break apps.

DeepDiver1975 committed:
Don't break app on travis

blizzz commented:
> All LDAP unit tests require DB access - even those who already mock IConfig etc :fish: -y

We're dealing with more tables than just those dealt with IConfig.

nickvergessen commented:
:+1:

MorrisJobke commented:
:+1:

oparoz commented:
> sure - but I'd take care of this in a PR of it's own - let's just make sure this PR does not break apps.

#fail. Breaks Gallery tests. There should have been a mailing list announcement to let devs know what they need to do to solve the new error message they may be getting.

nickvergessen commented:
Well we didn't break Travis, but yes should have sent an email nevertheless

oparoz commented:
Sure, Travis still works, but it's not a long term solution to bypass the check for one CI system only. People still test locally and might be using other CI solutions.

DeepDiver1975 commented:
> #fail. Breaks Gallery tests

can you please link the error? THX

nickvergessen commented:
@oparoz please see https://mailman.owncloud.org/pipermail/devel/2015-December/001810.html

oparoz commented:
@DeepDiver1975 - I can't link to it, you simply see the integration tests fail in the console because they need the DB, unlike unit tests.

@nickvergessen - Thanks for that. I should be able to modify the profile for those tests and get them to pass again.
